cpp/src/reshape/explode.cu

```diff
@@ -36,119 +36,225 @@
+                    counting_iter + gather_map.size(),
+                    gather_map.begin(),
+                    [position_array = pos.data(),
+                     offsets_minus_one,
+                     offsets,
+                     offset_size = explode_col.size()] __device__(auto idx) -> size_type {
+                      auto lb_idx =
+                        thrust::lower_bound(
+                          thrust::seq, offsets_minus_one, offsets_minus_one + offset_size, idx) -
+                        offsets_minus_one;
+                      position_array[idx] = idx - (offsets[lb_idx] - offsets[0]);
+                      return lb_idx;
+                    });
+
+  GatherType empty{0, stream};
+  return build_table(input_table,
+                     explode_column_idx,
+                     sliced_child,
+                     gather_map,
+                     empty,
+                     std::move(pos),
+                     true,
+                     stream,
+                     mr);
+}
+
+std::unique_ptr<table> explode_outer(table_view const& input_table,
+                                     size_type const explode_column_idx,
+                                     bool include_position,
+                                     rmm::cuda_stream_view stream,
+                                     rmm::mr::device_memory_resource* mr)
+{
+  lists_column_view explode_col{input_table.column(explode_column_idx)};
+  auto sliced_child  = explode_col.get_sliced_child(stream);
+  auto counting_iter = thrust::make_counting_iterator(0);
+  auto offsets       = explode_col.offsets_begin();
+
+  // number of nulls or empty lists found so far in the explode column
+  GatherType null_offset(explode_col.size(), stream);
+
+  auto null_iter = thrust::make_transform_iterator(
+    thrust::make_counting_iterator(0),
+    [offsets, offsets_size = explode_col.size() - 1] __device__(int idx) {
+      return (idx > offsets_size || (offsets[idx + 1] != offsets[idx])) ? 0 : 1;
+    });
+  thrust::inclusive_scan(
+    rmm::exec_policy(stream), null_iter, null_iter + sliced_child.size(), null_offset.begin());
+
+  auto null_or_empty_count = null_offset.size() > 0 ? null_offset.back_element(stream) : 0;
+  if (null_or_empty_count == 0) {
+    // performance penalty, run simple explode
+    return include_position ? explode_position(input_table, explode_column_idx, stream, mr)
+                            : explode(input_table, explode_column_idx, stream, mr);
+  }
+
+  auto gather_map_size = sliced_child.size() + null_or_empty_count;
+
+  GatherType gather_map(gather_map_size, stream);
+  GatherType explode_col_gather_map(gather_map_size, stream);
+  PositionType pos(include_position ? gather_map_size : 0, stream, mr);
+
+  // offsets + 1 here to skip the 0th offset, which removes a - 1 operation later.
+  auto offsets_minus_one = thrust::make_transform_iterator(
+    offsets + 1, [offsets] __device__(auto i) { return (i - offsets[0]) - 1; });
```

Comment:
Whenever you're advancing an iterator like this, it better expresses intent to use `next/prev`. Otherwise, when doing a quick read, it's hard to tell if you're doing arithmetic on a numeric value or manipulating an iterator. 
```suggestion
    thrust::next(offsets), [offsets] __device__(auto i) { return (i - offsets[0]) - 1; });
```